### ISSUE TITLE
Is there a way to define Deployed Apps memory setting ?

### ISSUE THREAD

**Issue by aamol** [spring-cloud/spring-cloud-dataflow#616]
Hi,
Whenever we create a new stream it spawns a new Java process for all its modules.
Even if the same module is repeated in multiple Stream they have a new Java Process for each time it is being used.
Please let me know if this is the actual behaviour or am I doing something wrong.

Also if this is the default behaviour is there are way to specify Xms and Xmx for these modules. Right now it is running default memory setting and gives me restriction on the number of Streams and Control over memory settings.

Thanks,
Amol

**Comment by sabbyanandan** [comment]
Thanks for trying Spring Cloud Data Flow (SCDF)!

All the apps in a stream are individual Spring Boot apps (_uber jar's_); hence, they run as as a separate Java process. 

As for the app specific memory override, we do provide that capability for platform specific SCDF implementations. For example, in SCDF's CF-server, you can override the memory setting via `spring.cloud.deployer.cloudfoundry.memory` and in SCDF's YARN-server, you can control the memory settings at the yarn-container and yarn-appmaster as following:

```
spring.cloud.deployer.yarn.app.streamappmaster.javaOpts=-Xms512m -Xmx512m
spring.cloud.deployer.yarn.app.streamcontainer.javaOpts=-Xms512m -Xmx512m
spring.cloud.deployer.yarn.app.taskappmaster.javaOpts=-Xms512m -Xmx512m
spring.cloud.deployer.yarn.app.taskcontainer.javaOpts=-Xms512m -Xmx512m
```

Lastly, questions like this can be better discussed in stackoverflow. Going forward, please ask any questions in that forum. GH is for reporting bugs and feature requests.

**Comment by pilb** [comment]
And is there a way using Server - Local?

**Comment by donovanmuller** [comment]
@pilb See this commit https://github.com/spring-cloud/spring-cloud-deployer/commit/afe19b08b0da399b735de75c6e4f051fc768d795

In my opinion, (and in the context of usage with Data Flow) you should be able to use the following [deployment properties](http://docs.spring.io/spring-cloud-dataflow/docs/1.1.0.BUILD-SNAPSHOT/reference/html/spring-cloud-dataflow-create-stream.html#_deployment_properties)

```
dataflow:>stream create --name ticktock --definition "time | log"
dataflow:>stream deploy ticktock --properties "app.time.JAVA_OPTS=-Xms256m,app.time.JAVA_OPTS=-Xmx256m"
```

However, this does not seem possible.

On further investigation, the above properties (`app.time.JAVA_OPTS`) do not even make it into the [`deploymentProperties`](https://github.com/spring-cloud/spring-cloud-deployer/blob/afe19b08b0da399b735de75c6e4f051fc768d795/spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/AbstractLocalDeployerSupport.java#L80) in the local deployer impl. This is because the deployment properties are pushed into the [`appProperties`](https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/DataFlowAppDefinition.java#L126)  in `DataFlowAppDefinition`and so is not included ([see above](https://github.com/spring-cloud/spring-cloud-deployer/blob/master/spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/AbstractLocalDeployerSupport.java#L80)).

Which means you *could* do this:

```
dataflow:>stream create --name ticktock --definition "time --JAVA_OPTS=-Xmx256m | log"
```

*if* the local deployer also checked `appProperties` on the deployment request app definition (it currently does not).

If you try the below in the hopes that it works similar to the YARN deployer:

```
dataflow:>stream deploy ticktock --properties "spring.cloud.deployer.app.time.JAVA_OPTS=-Xmx512m"
```

it also won't work because [`StreamDeploymentController#parseAndPopulateProperties`](https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentController.java#L296) only looks for deployment properties with an `app.` prefix.

If you had to include the app properties in `AbstractLocalDeployerSupport ` (see [this Gist](https://gist.github.com/donovanmuller/6c03dfaeccd21cc6e95663ddda0eb6af#file-abstractlocaldeployersupport-java-L19) for a basic example) then both the below options would work:

```
dataflow:>stream create --name ticktock --definition "time --JAVA_OPTS=-Xmx256m | log"

dataflow:>stream deploy ticktock --properties "app.time.JAVA_OPTS=-Xms256m,app.time.JAVA_OPTS=-Xmx256m"
```

So in short (and I could be missing something...), there is probably no good way to do this until the following issues are resolved:

* https://github.com/spring-cloud/spring-cloud-deployer/issues/116
* https://github.com/spring-cloud/spring-cloud-dataflow/issues/920

Also related to #966

**Comment by bpenergy** [comment]
Thank you for this answer.

Unfortunately, this is not working on 1.0.1.

I will wait tickets to be closed.

B.
